r2i/prediction.cc

```diff
@@ -0,0 +1,159 @@
+ * LLC.  No part of this program may be photocopied, reproduced or translated
+ * into another programming language without prior written consent of
+ * RidgeRun, LLC.  The user is free to modify the source code after obtaining
+ * a software license from RidgeRun.  All source code changes must be provided
+ * back to RidgeRun without any encumbrance.
+*/
+
+#include "r2i/prediction.h"
+
+#include <stdio.h>
+#include <string.h>
+
+namespace r2i {
+
+Prediction::~Prediction () {
+  this->results_data.clear();
+  this->results_sizes.clear();
+}
+
+double Prediction::At (unsigned int output_index, unsigned int index,
+                       r2i::RuntimeError &error) {
+  error.Clean ();
+
+  if (output_index >= this->results_data.size()) {
+    error.Set (RuntimeError::Code::MEMORY_ERROR, "Output index out of bounds");
+    return 0;
+  }
+
+  if (nullptr == this->results_data[output_index]
+      or 0 == this->results_sizes[output_index]) {
+    error.Set (RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER,
+               "Prediction was not properly initialized");
+    return 0;
+  }
+
+  unsigned int n_results =  this->GetResultSize(output_index, error);
+  if (RuntimeError::Code::EOK != error.GetCode()) {
+    return 0;
+  }
+  if (n_results < index ) {
+    error.Set (RuntimeError::Code::MEMORY_ERROR,
+               "Triying to access an non-existing index");
+    return 0;
+  }
+
+  float *fdata = static_cast<float *>(this->GetResultData(output_index, error));
+  if (RuntimeError::Code::EOK != error.GetCode()) {
+    return 0;
+  }
+  if (nullptr == fdata) {
+    error.Set (RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER,
+               "Prediction result not set yet");
+    return 0;
+  }
+
+  return fdata[index];
+}
+
+void *Prediction::GetResultData (unsigned int output_index,
+                                 RuntimeError &error) {
+  if (output_index >= this->results_data.size()) {
+    error.Set(RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER,
+              "Output index out of bounds");
+    return 0;
+  }
+
+  return this->results_data[output_index].get();
+}
+
+unsigned int Prediction::GetResultSize (unsigned int output_index,
+                                        RuntimeError &error) {
+  if (output_index >= this->results_sizes.size()) {
+    error.Set(RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER,
+              "Output index out of bounds");
+    return 0;
+  }
+
+  return this->results_sizes[output_index];
+}
+
+RuntimeError Prediction::AddResult (float *data, unsigned int size) {
+  RuntimeError error;
+
+  if (nullptr == data) {
+    error.Set (RuntimeError::Code::NULL_PARAMETER,
+               "Invalid tensor values passed to prediction");
+    return error;
+  }
+
+  if (0 == size) {
+    error.Set (RuntimeError::Code::NULL_PARAMETER,
+               "Invalid tensor size passed to prediction");
+    return error;
+  }
+
+  float *internal_data = (float *)malloc(size * sizeof(float));
```

Comment:
This is dangerous, I recommend to use shared_ptr instead. You can use this snippet below as an example of how to implement it:

`  this->output_data = std::shared_ptr<float>(new float[this->tensor_size],
                      std::default_delete<float[]>());
`

Reply:
I've reverted this change.